Add ignition2rpm code to mcbs-hackweek repo?

jkyros opened the pull request:
@cheesesashimi and I talked and I already handled the units/dropins and some of the mco/ostree logic so we decided to try and use this one. 

- I did not check in the vendored deps, but I will if you would like me to 
- I haven't typically nestled go code in other repos, so if you'd prefer I handle this a different way, let me know. :smiley: 
- This is hackweek quality code. **This is hackweek quality code.** 

( I figure eventaully Colin wants this to all work the CI, but I didn't mess with the build/test scripts as part of this )

sinnykumari commented:
Some overall suggestions:
Since this is adding high level go code, formatting the directory structure a bit better would help extend easily future go code.
   - go.sum, go.mod can go to top level directory
   -  ignition2rpm/cmd/main.go can be moved to cmd/ignition2rpm/main.go
   -  README can be either kept inside cmd/ignition2rpm/ for now and later can be moved to high level docs/ dir.
   - build and Makefile instruction can be merged into top level Makefile already present. And you can add a target to build ignition2rpm binary.

cheesesashimi commented:
Overall, this looks great! Just to capture some of my comments from the discussion @jkyros and I had yesterday:

- I like that this uses `rpmpack` directly. I used [NFPM](https://github.com/goreleaser/nfpm) (which also uses `rpmpack` under the hood) for [my implementation](https://github.com/cheesesashimi/mcbs-scratchspace/blob/main/cmd/mcfg2rpm/main.go). While NFPM has a nicer abstraction, it requires one to first write the contents to disk prior to packaging into an RPM. Using `rpmpack` directly avoids the need for this roundtrip.
- There are some improvements we can make (which we should address separately). For example, if you have more than one MachineConfig in the same file, this code as-is will only read the first MachineConfig. The rest will be ignored.

jkyros commented:
Ok cool. Thanks.  I didn't want to just take over the repo with go stuff without asking. :smile:  

I have: 
- moved cmd/ and the go.mod, go.sum to the root 
- updated the makefile with vendor and ignition2rpm targets 
- left the readme with the ignition2rpm command for now

jlebon commented on cmd/ignition2rpm/main.go:
```diff
@@ -0,0 +1,404 @@
+
+	flag.StringVar(&excludePrefix, "exclude-prefix", "", "Exclude files with this prefix")
+	flag.StringVar(&onceFrom, "config", "", "Config file ign/machineconfig to read")
+	flag.StringVar(&outputRPM, "output", "", "Specity name of RPM file to write")
```
```suggestion
	flag.StringVar(&outputRPM, "output", "", "Specify name of RPM file to write")
```

sinnykumari commented:
Thanks John for making the change.
lgtm, haven't looked closely into main.go logic that convert ignition content to rpm as they are in progress work :)

Will let others to review if there are other concerns.

jlebon commented on cmd/ignition2rpm/main.go:
```diff
@@ -0,0 +1,404 @@
+	flag.Set("v", "2")
+
+	flag.StringVar(&excludePrefix, "exclude-prefix", "", "Exclude files with this prefix")
+	flag.StringVar(&onceFrom, "config", "", "Config file ign/machineconfig to read")
```
Minor: would `config` be a better variable name for this?

jlebon commented on cmd/ignition2rpm/main.go:
```diff
@@ -0,0 +1,404 @@
+	hostname, _ := os.Hostname()
+
+	//specify some boilerplate metadata
+	packedRPM, err := rpmpack.NewRPM(rpmpack.RPMMetaData{
```
Oh wow, didn't know this existed.

jlebon commented on cmd/ignition2rpm/main.go:
```diff
@@ -0,0 +1,404 @@
+		packedRPM.RPMMetaData.Provides = append(packedRPM.RPMMetaData.Provides, &rpmpack.Relation{Name: "rpmostree(override)"})
+	}
+
+	if err != nil {
```
Shouldn't this be higher up?

cheesesashimi commented on cmd/ignition2rpm/README.md:
```diff
@@ -0,0 +1,24 @@
+Also works with MachineConfig ( or at least it seems to, but I haven't tested this exhaustively ). 
+
+Things to be aware of: 
+	- It relocates some files from /usr/local to /var/usrlocal to make rpm-ostree happy 
```
@jkyros Did we ever find a canonical source of these? It would probably be worth it to include a reference to it even as a comment.

jkyros replied:
Yes-ish? I found them in the rpm-ostree source: 

- The root ones: 
https://github.com/coreos/rpm-ostree/blob/d96748c353b3837124add5e6ebad083966533e04/rust/src/composepost.rs#L67
- Some other ones: 
https://github.com/coreos/rpm-ostree/blob/d96748c353b3837124add5e6ebad083966533e04/rust/src/composepost.rs#L749

cheesesashimi replied:
Cool! It's interesting that they have a special case for Vagrant: https://github.com/coreos/rpm-ostree/blob/d96748c353b3837124add5e6ebad083966533e04/rust/src/composepost.rs#L752.

cheesesashimi commented on cmd/ignition2rpm/main.go:
```diff
@@ -0,0 +1,404 @@
+	case mcfgv1.MachineConfig:
+
+		//shuck the ignition out of the machineconfig if need be
+		//TODO: fish the name out of the machineconfig for the metadata?
```
The MachineConfig has a `.GetName()` method attached to it that can provide this info.

jkyros replied:
Yeah, that TODO was more "I thought about using the machineconfig name, but I already made the RPM and set the metadata up there on line 68 based on the filename, so that might be a cool idea...later...if we keep developing it"   :smile:

cheesesashimi replied:
No worries! I mentioned it yesterday and somehow forgot to include it as one of my top-level comments.

cheesesashimi commented on cmd/ignition2rpm/main.go:
```diff
@@ -0,0 +1,404 @@
+	}
+
+	// Try to parse as a machine config
+	mc, err := resourceread.ReadMachineConfigV1(content)
```
I mentioned this elsewhere, but I'll also mention it here for posterity: This function will only return the first MachineConfig it finds in the byte array. Meaning that if you have multiple MachineConfigs within a single YAML file, only the first one will be read. The rest will be ignored.

EDIT: To be clear, I don't think this is required for merging this. A comment indicating that will suffice for now.

jlebon reviewed (approved):
Didn't review everything, but looks sane overall! One thing that would be very convincing is adding a test Ignition config and a test MachineConfig at least to play with, even if we hook it up to CI in a follow-up.

jkyros committed:
Add igniton2rpm to mcbs-hackweek repo

This adds code to generate rpms from ignition and machine config
to mcbs-hackweek, so we can use it to generate rpms that cen be
applied by the "override builder"

jkyros commented:
Fixed all the things Jonathan noticed. Added the code to actually decode the files now -- kinda useful. :smiley:

cgwalters reviewed (approved):
Let's go forward with this!  Nice work
